docs/auth.md

```diff
@@ -0,0 +1,462 @@
+against stealing authorization code with redirection, the fragment is always
+overriden:
+
+```http
+GET /auth/login?redirect=https://contacts.cozy.example.org/foo?bar#baz HTTP/1.1
+Host: cozy.example.org
+Cookies: ...
+```
+
+**Note**: the redirect parameter should be URL-encoded. We haven't done that
+to make it clear what the path (`foo`), the query-string (`bar`), and the
+fragment (`baz`) are.
+
+```http
+HTTP/1.1 302 Moved Temporarily
+Location: https://contacts.cozy.example.org/foo?bar#_=_
+```
+
+If the `redirect` parameter is invalid, the response will be `400 Bad
+Request`. Same for other parameters, the redirection will happen only on
+success (even if OAuth2 says the authorization server can redirect on errors,
+it's very complicated to do it safely, and it is better to avoid this trap).
+
+### POST /auth/login
+
+After the user has typed her password and clicked on `Login`, a request is
+made to this endpoint.
+
+The `redirect` parameter is passed inside the body. If it is missing, the
+redirection will be made against the default target: the home application of
+this cozy instance.
+
+```http
+POST /auth/login HTTP/1.1
+Host: cozy.example.org
+Content-type: application/x-www-form-urlencoded
+
+password=p4ssw0rd&redirect=https%3A%2F%2Fcontacts.cozy.example.org
+```
+
+```http
+HTTP/1.1 302 Moved Temporarily
+Set-Cookie: ...
+Location: https://contacts.cozy.example.org/foo
+```
+
+### DELETE /auth/login
+
+This can be used to log-out the user. A private context must be passed in the
+query-string, to protect against CSRF attack on this (this can part of bigger
+attacks like session fixation).
+
+```http
+DELETE /auth/login?CtxToken=token-for-a-private-context HTTP/1.1
+Host: cozy.example.org
+```
+
+### POST /auth/register
+
+This route is used by OAuth2 clients to dynamically register them-selves.
+
+See [OAuth 2.0 Dynamic Client Registration
+Protocol](https://tools.ietf.org/html/rfc7591) for the details.
+
+It gives to the client these informations:
+
+- `client_id`
+- `client_secret`
+- `registration_access_token`
+
+### GET /auth/register/:client-id
+### PUT /auth/register/:client-id
+### DELETE /auth/register/:client-id
+
+These routes follow the [OAuth 2.0 Dynamic Client Registration Management
+Protocol RFC](https://tools.ietf.org/html/rfc7592). They allow an OAuth2
+client to get back its metadata, update them, and unregister itself.
+
+The client has to sent its registration access token to be able to use this
+endpoint.
+
+### GET /auth/authorize
+
+When an OAuth2 client wants to get access to the data of the cozy owner, it
+starts the OAuth2 dance with this step. The user is shown what the client asks
+and has an accept button if she is OK with that.
+
+The parameters are:
+
+- `client_id`, that identify the client
+- `redirect_uri`, it has be exactly the same of the one used in registration
+- `state`, it's a protection against CSRF on the client (a random string
```

Comment:
If the fragment is overwritten and the user use a link & exit the apps page, the state is not only a random string but also contains the current state of the client if it is a SPA  / a sessionid if it's not.